### PR TITLE
Add postgresql based persistence class

### PR DESCRIPTION
**Pull request by starry69**
Adds custom persistence class derived from `ext.BasePersistence` using Postgresql database as backend.

### CODE REVIEW

**Review by newblaque — APPROVED** [review]

**Comment by Bibo-Joshi** [comment]
Hi! Sorry for the very late reply … First: thanks for the PR! I have a few notes:

* Pre-commit and Codacy are failing, please look into that
* While I'm sure that you tested the persistence class manually, we will need at least some tests for the automated test setup. Please have a look at the tests of roles on how to install the additional dependencies and maybe at the persistence tests of ptb for general inspiration (those are pretty elaborate, you don't have to do tests to that extend …). Feel free to reach out on telegram if you have questions on tests :)
* It would be good to document in the readme what kind of data can be handled by this persistence class (iisc data needs to be json-dumpable) and maybe also that data will be written to the DB only on shutdown.
* Just a thought: If it's correct that only json-dumpable data can be handled, it may be beneficial to base this class on `DictPersistence`. IISC that way you'd only have to implement `flush` and the `__init__`.

**Comment by starry69** [comment]
Hi! I totally agree that it should be inherited from `DictPersistence` instead of `BasePersistence`, dunno why i haven't thought about it at first. I get why pre-commit was failing but couldn't figure out codacy part.. it says something about readme which i didn't understand well enough.. tho will see again while adding further documentation about flush and type of data it handles.

File: ptbcontrib/postgres_persistence/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Persistence class with Postgresql database as backend.
```
Comment (Poolitzer) [inline ptbcontrib/postgres_persistence/README.md:1]: I think codacy is actually pretty clear this time. here it complains about the trailing dot at headings.
```suggestion
# Persistence class with Postgresql database as backend
```

File: ptbcontrib/postgres_persistence/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+* How to use:
```
Comment (Poolitzer) [inline ptbcontrib/postgres_persistence/README.md:3]: And here that you do a list item without indenting it. Id suggest making it a level 2 header.
```suggestion
## How to use:
```

File: ptbcontrib/postgres_persistence/postgrespersistence.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    def __init_database(self) -> None:
+        """creates table for persistence data if not exists"""
+
+        create_table_qry = """
+            CREATE TABLE IF NOT EXISTS persistence(
+            data json NOT NULL);"""
+        self._session.execute(text(create_table_qry))
+        self._session.commit()
+
+    def __load_database(self) -> None:
```
Comment (Bibo-Joshi) [inline ptbcontrib/postgres_persistence/postgrespersistence.py:91]: Didn't try to understand in detail what's happening in here, so again just a thought: `DictPersistence` can accept json-string to recover the data from on init. Would loading those strings from DB and passing it to the parents `__init__` work?

File: tests/test_psqlpersistence.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+        def second(update, context):
+            if not context.user_data['test1'] == 'test2':
+                pytest.fail()
+            if not context.chat_data[3] == 'test4':
+                pytest.fail()
+            if not context.bot_data['test1'] == 'test2':
+                pytest.fail()
+
+        h1 = MessageHandler(None, first, pass_user_data=True, pass_chat_data=True)
```
Comment (Bibo-Joshi) [inline tests/test_psqlpersistence.py:114]: `pass_*` is depercated since v12, you can skip that.

File: tests/test_psqlpersistence.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        dp.remove_handler(h1)
+        dp.add_handler(h2)
+        dp.process_update(update)
+
+        assert self.executed != ""
+        assert self.commited == 555
+        assert self.ses_closed is True
+
+    def test_flush(self, bot, update, monkeypatch):
```
Comment (Bibo-Joshi) [inline tests/test_psqlpersistence.py:127]: could you also add a test that checks that loading from on start works fine?

**Review by Bibo-Joshi — COMMENTED** [review]
Thanks for the updates! Left a few more comments. Regarding `pre-commit`: the setup of that tests only installs the general dependencies. So I guess we should just ignore import-erros (same problem will happen for other contributions that have 3rd party dependencies). Could you add that to setup.cfg? see ptbs [setup.cfg](https://github.com/python-telegram-bot/python-telegram-bot/blob/80b34811ab7cfb633d703f88d2a06df6e71d3551/setup.cfg#L21). As you see, ptbcontrib is still very young and the setup problems appear as we go :D

File: ptbcontrib/postgres_persistence/postgrespersistence.py (hunk 3 above)
Comment (starry69) [reply ptbcontrib/postgres_persistence/postgrespersistence.py:91]: I still need to divide it into separate json strings, i.e for `user/chat bot_data` and conversation data, as json is saved into database as a single json string by using single row only without any primary key.. so ig it'll be better if we leave it as it is.

Comment (starry69) [reply ptbcontrib/postgres_persistence/postgrespersistence.py:91]: Adding primary key setup will be mess as postgresql is rational database so we'll need to make 4 different columns then save then using 4 unique primary keys

File: tests/test_psqlpersistence.py (hunk 5 above)
Comment (starry69) [reply tests/test_psqlpersistence.py:127]: Sure

File: tests/test_psqlpersistence.py (hunk 4 above)
Comment (starry69) [reply tests/test_psqlpersistence.py:114]: 👍

File: ptbcontrib/postgres_persistence/postgrespersistence.py (hunk 3 above)
Comment (Bibo-Joshi) [reply ptbcontrib/postgres_persistence/postgrespersistence.py:91]: > I still need to divide it into separate json strings, i.e for `user/chat bot_data` and conversation data, as json is saved into database as a single json string by using single row only, without any primary key.. so ig it'll be better if we leave it as it is.

yeah. okay. was just a thought ;)

File: setup.cfg [hunk 6]
```diff
@@ -4,6 +4,9 @@ license_file = LICENSE
 [flake8]
 max-line-length = 99
 
+[pylint.message-control]
+disable = C0330,R0801,R0913,R0904,R0903,R0902,W0511,C0116,C0115,W0703,R0914,R0914,C0302,R0912,R0915,R0401
```
Comment (Bibo-Joshi) [inline setup.cfg:8]:
```suggestion
disable = E0401
```

This disables the `import-error`, while leaving all the other warnings in tact ;)

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 6 above)
Comment (starry69) [reply setup.cfg:8]: Oops!

**Comment by Bibo-Joshi** [comment]
Thank you for the contribution! :)

**Pull request merged**